rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,526 @@
+so we would probably have to leverage the ‘main worker’ concept here.
+
+The SO migration v2 is supposed to be resilient to concurrent attempts though, as we already support multi-instances 
+Kibana, so this can probably be considered an improvement.
+
+## Open questions and things to solve
+
+### Memory consumption
+
+In clustered mode, node options such as `max-old-space-size` will be used by all processes. 
+
+E.g using `--max-old-space-size=1024` in a 2 workers cluster would have a maximum memory usage of 3gb (1 coordinator + 2 workers). 
+
+If this something we will need to document somewhere?
+
+### Workers error handling
+
+When using `cluster`, the common best practice is to have the coordinator recreate ('restart') workers when they terminate unexpectedly. 
+However, given Kibana's architecture, some failures are not recoverable (workers failing because of config validation, failed migration...). 
+
+For instance, if a worker (well, all workers) terminates because of an invalid configuration property, it doesn't make
+any sense to have the coordinator recreate them indefinitely, as the error requires manual intervention. 
+
+Should we try to distinguish recoverable and non-recoverable errors, or are we good terminating the main Kibana process 
+when any worker terminates unexpectedly for any reason (After all, this is already the behavior in non-cluster mode)?
+
+### Data folder
+
+The data folder (`path.data`) is currently the same for all workers. 
+
+We still have to identify with the teams if this is going to be a problem. It could be, for example, if some plugins
+are accessing files in write mode, which could result in concurrency issues between the workers.
+
+If that was confirmed, we would have to create and use a distinct data folder for each worker.
+
+One pragmatic solution could be, when clustering is enabled, to create a sub folder under path.data for each worker. 
+
+If we need to do so, should this be considered a breaking change, or do we assume the usage of the data folder is an 
+implementation detail and not part of our public 'API'?
+
+### instanceUUID
+
+The same instance UUID (`server.uuid` / `{dataFolder}/uuid`) is currently used by all the workers. 
+
+We still need to identify with the teams if this is going to be a problem, in which case we would need to have distinct 
+instance uuid per workers.
+
+Note that this could be a breaking change, as the single `server.uuid` configuration property would not be enough. 
+Could the new `clustering.getWorkerId()` API be used here somehow? We could have a unique worker id with `serverUUid-workerId`. 
+
+### The Dev CLI
+
+In development mode, we are already spawning processes from the main process: The Kibana server running in the main process 
+actually just kickstarts core to get the configuration and services required to instantiate the `CliDevMode` 
+(from the legacy service), which itself spawns a child process to run the actual Kibana server.
```

Comment:
@mistic can Bazel be of any aid with this? I think that once Bazel is handling the webpack builds, the only thing we really need multiple processes for in the dev CLI is for base path proxying. It'd be neat if we could create a Bazel target that starts up an nginx instance for base path proxying and a Kibana server process (w/ clustering enabled if desired).

Reply:
This will be addressed by https://github.com/elastic/kibana/issues/76935, which will very likely land before any implementation attempt for this proposal. I wonder if I should just remove this section?

Reply:
removed